src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -161,7 +182,9 @@ private void AllocateBuffer()
             //   and start reading into _current. Set _index to 0.
             // - When _current runs out of space, add it to _buffers and repeat the
             //   above step, except with _current.Length * 2.
+            // - Make sure we never pass _maxCapacity in all of the above steps.
 
+            Debug.Assert((uint)_maxCapacity > (uint)_count);
```

Comment:
I think we may need to keep the uint cast here. We could have come here from `AddRange`, where we set `_count` to int.MinValue (last buffer is 0x40000000). If that were the case and we removed these casts, the assert will pass even though we've exceeded our limit. It's different from `Add` where we didn't modify that field yet.